### PR TITLE
lgc: Accept num patch control points via shader mode

### PR DESCRIPTION
**Pull request by trenouf**

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5c911a3f57764cc17a872ea5b4ea3a2ba6e5e5c9
<details><summary>CTS tests (Failed: 0/208297)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36986/69439 (53.3%)</li>
<li>Failed: 0/69439 (0.0%)</li>
<li>Not Supported: 32453/69439 (46.7%)</li><li>Warnings: 0/69439 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37206/69417 (53.6%)</li>
<li>Failed: 0/69417 (0.0%)</li>
<li>Not Supported: 32211/69417 (46.4%)</li><li>Warnings: 0/69417 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 34225/69441 (49.3%)</li>
<li>Failed: 0/69441 (0.0%)</li>
<li>Not Supported: 35216/69441 (50.7%)</li><li>Warnings: 0/69441 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
Why support both the ShaderMode and the InputAssemblyState? Wouldn't it be better to only support one?

**Comment by trenouf** [comment]
I changed it so LGC only supports ShaderMode, but it's a bit shonky with the hoops that the front-end has to jump through, especially with the need to include patchControlPoints in the part pipeline state hash. I think the original scheme is better where the front-end can use either ShaderMode or InputAssemblyState for patchControlPoints.

**Comment by trenouf** [comment]
Ping @nhaehnle -- can I go back to my original scheme (modulo the two fixes I did for it)? Or do you still prefer the only-ShaderMode scheme?

File: llpc/context/llpcPipelineContext.h [hunk 1]
```diff
@@ -146,6 +146,9 @@ class PipelineContext {
   // Set pipeline state in lgc::Pipeline object for middle-end, and (optionally) hash the state.
   virtual void setPipelineState(lgc::Pipeline *pipeline, Util::MetroHash64 *hasher, bool unlinked) const;
 
+  // For TCS, set inputVertices from patchControlPoints in the pipeline state.
+  virtual void setTcsInputVertices(llvm::Module *tcsModule) {}
+
```
Comment (nhaehnle) [inline llpc/context/llpcPipelineContext.h:151]: Maybe put an `llvm_unreachable` in there just in case?

**Review by nhaehnle — COMMENTED** [review]
I do think it's best to have the option only in one place.

Yes, the `hasher` thing is a bit unfortunate but it's always been a bit strange in the first place to have a "setPipelineState" method optionally compute a hash. That's also only used in one place.

**Comment by trenouf** [comment]
OK, I'll go with your way, as I need to get the change in. I still think it is the wrong way to do it as it spreads shonkiness around the front-end.

setPipelineState computing the hash is deliberate because, for the part pipeline scheme, we need to make sure the hash includes LGC pipeline state needed for that part pipeline, without risking someone adding some new state and forgetting to include it in the hash. It was less strange before this change.

**Review by nhaehnle — APPROVED** [review]
Thanks.

And yeah, the main part that makes me uncomfortable about the hashing is that it's *only* used for part pipelines. That feels a bit arbitrary. But it's also not the point here.

**Pull request merged**